### PR TITLE
Revise bounds check for MCL

### PR DESCRIPTION
**Pull request by bennibolm**
As for the IDP limiting, the bounds check now also exports the maximum deviation since the last export.

### CODE REVIEW

File: examples/tree_2d_dgsem/elixir_euler_sedov_blast_wave_MCL.jl [hunk 1]
```diff
@@ -94,7 +94,8 @@ callbacks = CallbackSet(summary_callback,
 ###############################################################################
 # run the simulation
 
-stage_callbacks = (BoundsCheckCallback(save_errors=false),)
+output_directory = "out"
+stage_callbacks = (BoundsCheckCallback(save_errors=true, interval=100, output_directory = output_directory),)
```
Comment (amrueda) [inline examples/tree_2d_dgsem/elixir_euler_sedov_blast_wave_MCL.jl]: I believe "out" should be the default output directory. Why is this needed in the elixir?
```suggestion
stage_callbacks = (BoundsCheckCallback(save_errors=true, interval=100),)
```

File: src/solvers/dgsem_tree/subcell_limiters_2d.jl [hunk 2]
```diff
@@ -725,7 +725,9 @@ function create_cache(limiter::Type{SubcellLimiterMCL}, equations::AbstractEquat
                                                                  nvariables(equations),
                                                                  nnodes(basis))
 
-    idp_bounds_delta = zeros(real(basis), 2,
+    # Memory for bounds checking routine with `BoundsCheckCallback`.
+    # [maximum since the last export / total maximum, min / max, variable]
+    idp_bounds_delta = zeros(real(basis), 2, 2,
                              nvariables(equations) + PressurePositivityLimiterKuzmin)
```
Comment (amrueda) [inline src/solvers/dgsem_tree/subcell_limiters_2d.jl]: It would be good to rename the limiters to lower-case (e.g., `PressurePositivityLimiterKuzmin` -> `pressure_positivity_limiter_kuzmin` or similar). But this can also be done in another PR. BTW, do you have the sharp and the cautious variants of the limiter in this implementation?

File: src/solvers/dgsem_tree/subcell_limiters_2d.jl [hunk 3]
```diff
@@ -725,7 +725,9 @@ function create_cache(limiter::Type{SubcellLimiterMCL}, equations::AbstractEquat
                                                                  nvariables(equations),
                                                                  nnodes(basis))
 
-    idp_bounds_delta = zeros(real(basis), 2,
+    # Memory for bounds checking routine with `BoundsCheckCallback`.
+    # [maximum since the last export / total maximum, min / max, variable]
+    idp_bounds_delta = zeros(real(basis), 2, 2,
```
Comment (amrueda) [inline src/solvers/dgsem_tree/subcell_limiters_2d.jl]: Is there a reason why you use the name `idp_bounds_delta` for MCL?

File: test/test_tree_2d_euler.jl [hunk 4]
```diff
@@ -321,12 +321,16 @@ EXAMPLES_DIR = pkgdir(Trixi, "examples", "tree_2d_dgsem")
...
+    rm("out/deviations.txt", force=true)
     @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_sedov_blast_wave_MCL.jl"),
       l2   = [0.4740321851943766, 0.15889871334104985, 0.15889871334104988, 0.6190405536267991],
       linf = [4.011954283668753, 1.8527131099524292, 1.8527131099524277, 6.465833729130187],
       tspan = (0.0, 1.0),
       initial_refinement_level=4,
       coverage_override = (maxiters=6,))
+      lines = readlines("out/deviations.txt")
+      @test lines[1] == "# iter, simu_time, rho_min, rho_max, rho_v1_min, rho_v1_max, rho_v2_min, rho_v2_max, rho_e_min, rho_e_max, pressure_min"
+      @test startswith(lines[end], "349") || startswith(lines[end], "1")
```
Comment (amrueda) [inline test/test_tree_2d_euler.jl]: What exactly do you test here? that the number of lines of the file is 349 or 1?

**Review by amrueda — COMMENTED** [review]
In general looks good to me. A couple of minor comments below.

File: src/solvers/dgsem_tree/subcell_limiters_2d.jl (hunk 2 above)
Comment (bennibolm) [reply src/solvers/dgsem_tree/subcell_limiters_2d.jl]: Yeah, this renaming issue is on my TODO list. I wanted to do it all together.

Comment (bennibolm) [reply src/solvers/dgsem_tree/subcell_limiters_2d.jl]: Yes, we should have this. At least the parameter `PressurePositivityLimiterKuzminExact` exists.

File: src/solvers/dgsem_tree/subcell_limiters_2d.jl (hunk 3 above)
Comment (bennibolm) [reply src/solvers/dgsem_tree/subcell_limiters_2d.jl]: No, I just copied it from the IDP limiting. What name would be better? Just `mcl_bounds_delta`?

Comment (amrueda) [reply src/solvers/dgsem_tree/subcell_limiters_2d.jl]: Yeah, maybe that's better

File: test/test_tree_2d_euler.jl (hunk 4 above)
Comment (bennibolm) [reply test/test_tree_2d_euler.jl]: In general, I want to test if the creation of the deviation file worked.
First, I just wanted to test if the amount of time steps (`349`) is correct. The problem was that the tests with coverage in general use `maxiters=1` or in this case `maxiters=6`. Therefore, I have to allow both cases.
In the already existing test for idp (see [here](https://github.com/trixi-framework/Trixi.jl/blob/dc0dc1d220e79d31465818acae235adf69ecddea/test/test_tree_2d_eulermulti.jl#L38-L40)) the amount of time steps is `15`. Michael suggested to just check for `1` to cover both cases. That doesn't work here. 

And I still check for `1` and not `6` since `interval=100>5` and the last step in an interrupted simulation is not known as the `finalstep`. Therefore, no deviation is exported.

File: examples/tree_2d_dgsem/elixir_euler_sedov_blast_wave_MCL.jl (hunk 1 above)
Comment (bennibolm) [reply examples/tree_2d_dgsem/elixir_euler_sedov_blast_wave_MCL.jl]: I added it for the case that someone adjusts `output_directory` in the elixir file. Then, we would save our file to this location and the check fails since `deviation.txt` is not found.
Like I implemented it, for the test, `output_directory` from the elixir is always overwritten and set to "out".

Comment (bennibolm) [reply examples/tree_2d_dgsem/elixir_euler_sedov_blast_wave_MCL.jl]: But I agree with you that it's a bit weird. If someone adjusts `save_errors=false` or just sets `interval` to `5` or smaller, the test fails anyway.

File: src/solvers/dgsem_tree/subcell_limiters_2d.jl (hunk 3 above)
Comment (bennibolm) [reply src/solvers/dgsem_tree/subcell_limiters_2d.jl]: Done in [add1d89](https://github.com/bennibolm/Trixi.jl/pull/118/commits/add1d89e9741b1952babc38951b152faca12f2ba)

File: examples/tree_2d_dgsem/elixir_euler_sedov_blast_wave_MCL.jl (hunk 1 above)
Comment (amrueda) [reply examples/tree_2d_dgsem/elixir_euler_sedov_blast_wave_MCL.jl]: ok!

File: src/callbacks_stage/subcell_bounds_check.jl [hunk 5]
```diff
@@ -128,7 +126,7 @@ function init_callback(callback::BoundsCheckCallback, semi, limiter::SubcellLimi
         if limiter.PressurePositivityLimiterKuzmin
             print(f, ", pressure_min")
         end
-        # No check for entropy limiting rn
+        # TODO: Bounds check for entropy limiting
```
Comment (amrueda) [inline src/callbacks_stage/subcell_bounds_check.jl:129]: Maybe it would be good to show a warning to the user saying that there is no bounds check for the entropy limiter.

**Review by amrueda — COMMENTED** [review]
Looks great to me!
Just a minor comment:

File: src/callbacks_stage/subcell_bounds_check.jl (hunk 5 above)
Comment (bennibolm) [reply src/callbacks_stage/subcell_bounds_check.jl:129]: Thanks. That's a good point. I added a warning to the console output in [a83951e](https://github.com/bennibolm/Trixi.jl/pull/118/commits/a83951e79d11d240af7e6f650e6e68c2ff43b2ce).
